npm/qsharp/src/compiler/compiler.ts

```diff
@@ -74,11 +75,18 @@ export interface ICompiler {
  * Type definition for the configuration of a program.
  * If adding new properties, make them optional to maintain backward compatibility.
  */
-export type ProgramConfig = {
-  /** An array of source objects, each containing a name and contents. */
-  sources: [string, string][];
-  /** An array of language features to be opted in to in this compilation. */
-  languageFeatures?: string[];
+export type ProgramConfig = (
```

Comment:
Well, I'd describe it a bit more differently, and I'd say the type _is_ fulfilling its job. It:

- Encapsulates all the program-related parameters so we don't have to add backcompat overloads or _every single function_ that takes a program. We just need to update this type
- Gives us the flexibility to reorder parameters by making everything named (e.g. so we don't end up with code like `compile(undefined, undefined, someparam)` in QCOM

Also, I don't really consider single-file mode "legacy" - for example, the Playground will be single file mode for the foreseeable future. I think there's value in allowing a simpler interface where you can pass in sources directly like this.

I honestly think this type is fine, and it might be TypeScript itself that's bothering you? It's an input object that can either have `sources` or `packageGraphSources` and I think that's just fine.